[linux-0.77.0] ? Disabled resizable window, only compatible with OpenGL output, & "windowresolution outside of desktop"

### Linux
https://github.com/dosbox-staging/dosbox-staging/issues/1129

at my first testing of 0.77.0 i was taken by the output forced to 640x480 windowed mode
i am used to my windowed mode taking up a 4:3 ratio of my screen at 1400x1050 (because reasons)
two monitors at 1920x1080, one with 1920x1024 surface area and the other with 1920x1080 area
i pulled out the `config -wcd` to check whats different

nothing (critically) changed, thats good
```
output           = texturenb
texture_renderer = opengl
```

texturenb, opengl.  as it was
so trying some settings like openglpp, openglnb, etc  now its declaring that i cant "scale" the window to my desired height and forces it down to 1280x960 window
or it complains that my second monitor is only 1920x1024, which its not.  thats the main monitor area, not even the area or monitor that it opens onto.

```
MAIN: windowresolution '1400x1050' outside of desktop '1920x1024', using 640x480 instead
MAIN: Sized window to 640x480 with nearest-neighbour scaling and stretched pixels

```

```
MAIN: Disabled resizable window, only compatible with OpenGL output
MAIN: Sized window to 1280x960 with nearest-neighbour scaling and stretched pixels

```

while in openglpp then going full screen its even worse,  the screen is totally wrong
wait so its forcing 640x400 no matter what in fullscreen mode?
where is the right side of the screen?  its missing entirely.  the left is all black border?
full screen means widescreen area inside a 4:3 surface?

```
MAIN: Disabled resizable window, only compatible with OpenGL output
MAIN: Sized window 960x800 with pixel-perfect scaling and stretched pixels
MAIN: Pixel-perfect scaling (2x2): 640x480 (PAR 1.00) -> 1280x960 (PAR 1.00)
SDL: Using driver "opengl" for texture renderer
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
SDL: Using driver "opengl" for texture renderer

```

???

id like my window to be 1400x1050 at all times even if its destroying the pixel aspect!
this is how i have all my dosbox forks setup.
when i put in the config
`fullresolution   = 1440x1080`
`windowresolution = 1400x1050`
thats exactly what i expect the output size to be

must be user error, not understanding the new changes in settings, wrestling with a configuration that doesnt accept my preferences?
this may seem like a lot of ranting and disorganization, im tired, i gave 0.77 no less than 25 minutes of my time to "test" the new release and settings changes..
i know, you want photos.  documentation of my settings compared to the output
_i will do that later_
if that is even relevant


**i love the idea of HW PP aspect and agree its the way forward**
making it so that i cant choose an alternative resolution really made me symlink 0.76.0 and move backwards for now

compiling this still only detects the 'main' monitor area, not the area that its expected to open in
*desktop*

```
windowresolution = desktop
output           = openglnb
texture_renderer = auto

```

`MAIN: Sized window to 1366x1024 with nearest-neighbour scaling and stretched pixels
`

its detecting 1920x1024 (left) as the area size. it should detect the full 1920x1080 (right) without top & bottom panel


setting to

```
windowresolution = 1400x1050
output           = openglnb
texture_renderer = auto

```

```
MAIN: Requested windowresolution '1400x1050' is larger than the desktop '1920x1024'
MAIN: Sized window to 1400x1050 with nearest-neighbour scaling and stretched pixels

```

**great that works as expected**

the other issue is openglpp changing monitor resolution

```
windowresolution = desktop <-- setting irrelevant
output           = openglpp
texture_renderer = auto

```

```
MAIN: Sized window to 1280x1000 with pixel-perfect scaling and stretched pixels
OPENGL: Pixel buffer object extension: available
MAIN: Pixel-perfect scaling (2x2): 640x480 (PAR 1.00) -> 1280x960 (PAR 1.00)
##alt+enter fullscreen
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20

```

~[.](https://i.imgur.com/s6dTvZC.png) 1280x800? or 640x400?
thats exactly the size of the window,  border on the left, missing area on the right

running Doom the monitor changes resolution again, PP is correct
exiting doom the fullscreen is now showing the full resolution centered correctly

other: opengl

`output = opengl
`

```
MAIN: Requested windowresolution '1400x1050' is larger than the desktop '1920x1024'
MAIN: Sized window to 1400x1050 with stretched pixels
OPENGL: Pixel buffer object extension: available
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
```



**NEXT *the build error***
python3.6 vs python3.8
note most of my system does not work correctly with python3.7, python3.8

> meson setup -Dbuildtype=release build

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 574, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 892, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 783, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (meson 0.53.1 (/usr/local/lib/python3.6/dist-packages), Requirement.parse('meson==0.53.2'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/meson", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3088, in <module>
    @_call_aside
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3072, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 3101, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 576, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 589, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 778, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'meson==0.53.2' distribution was not found and is required by the application

```
>sudo update-alternatives --config python3

```
There are 3 choices for the alternative python3 (providing /usr/bin/python3).

  Selection    Path                Priority   Status
------------------------------------------------------------
  0            /usr/bin/python3.6   1         auto mode
* 1            /usr/bin/python3.6   1         manual mode
  2            /usr/bin/python3.7   1         manual mode
  3            /usr/bin/python3.8   1         manual mode

Press <enter> to keep the current choice[*], or type selection number: 3
update-alternatives: using /usr/bin/python3.8 to provide /usr/bin/python3 (python3) in manual mode

```

the fluidsynth failure
yes i have fluidsynth-dev installed

> ninja -C build

```
[204/206] Compiling C++ object 'src/midi/08c63a5@@midi@sta/midi_fluidsynth.cpp.o'.
FAILED: src/midi/08c63a5@@midi@sta/midi_fluidsynth.cpp.o 
ccache c++ -Isrc/midi/08c63a5@@midi@sta -Isrc/midi -I../src/midi -I../include -I. -I../ -Isubprojects/munt-libmt32emu_2_5_0/include -I../subprojects/munt-libmt32emu_2_5_0/include -I/usr/include/SDL2 -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -O3 -Weffc++ -fPIC -D_REENTRANT -MD -MQ 'src/midi/08c63a5@@midi@sta/midi_fluidsynth.cpp.o' -MF 'src/midi/08c63a5@@midi@sta/midi_fluidsynth.cpp.o.d' -o 'src/midi/08c63a5@@midi@sta/midi_fluidsynth.cpp.o' -c ../src/midi/midi_fluidsynth.cpp
../src/midi/midi_fluidsynth.cpp: In member function ‘virtual void MidiHandlerFluidsynth::PlayMsg(const uint8_t*)’:
../src/midi/midi_fluidsynth.cpp:365:3: error: ‘fluid_synth_key_pressure’ was not declared in this scope
   fluid_synth_key_pressure(synth.get(), chanID, msg[1], msg[2]);
   ^~~~~~~~~~~~~~~~~~~~~~~~
../src/midi/midi_fluidsynth.cpp:365:3: note: suggested alternative: ‘fluid_synth_channel_pressure’
   fluid_synth_key_pressure(synth.get(), chanID, msg[1], msg[2]);
   ^~~~~~~~~~~~~~~~~~~~~~~~
   fluid_synth_channel_pressure
ninja: build stopped: subcommand failed.
```

>meson --reconfigure -Duse_fluidsynth=false -Dbuildtype=release build

build completed (which is how i did further testing with `dont-adjust-WxH-1.zip` )

--- a/src/gui/sdlmain.cpp
+++ b/src/gui/sdlmain.cpp
@@ -2143,8 +2143,6 @@ static void setup_window_sizes_from_conf(const char *windowresolution_val,
                                          const SCALING_MODE scaling_mode,
                                          const bool wants_stretched_pixels)
 {
-	assert(sdl.display_number >= 0);
-
 	// TODO: Deprecate SURFACE output and remove this.
 	// For now, let the DOS-side determine the window's resolution.
 	if (sdl.desktop.want_type == SCREEN_SURFACE)
@@ -2152,9 +2150,10 @@ static void setup_window_sizes_from_conf(const char *windowresolution_val,
 
 	sdl.desktop.want_resizable_window = detect_resizable_window();
 
-	// Get the total usable desktop resolution
+	// Get the total desktop resolution
 	SDL_Rect desktop;
-	SDL_GetDisplayUsableBounds(sdl.display_number, &desktop);
+	assert(sdl.display_number >= 0);
+	SDL_GetDisplayBounds(sdl.display_number, &desktop);
 	assert(desktop.w >= FALLBACK_WINDOW_DIMENSIONS.x);
 	assert(desktop.h >= FALLBACK_WINDOW_DIMENSIONS.y);
 

--- a/src/gui/sdlmain.cpp
+++ b/src/gui/sdlmain.cpp
@@ -1994,7 +1994,7 @@ static SDL_Point remove_stretched_aspect(const SDL_Point &size)
 	return {size.x, ceil_sdivide(size.y * 5, 6)};
 }
 
-static SDL_Point refine_window_size_with_mode(const SDL_Point &size,
+static SDL_Point refine_window_size(const SDL_Point &size,
                                               const SCALING_MODE scaling_mode,
                                               const bool wants_stretched_pixels)
 {
@@ -2112,22 +2112,27 @@ static SDL_Point window_bounds_from_label(const std::string &pref,
 }
 
 // Takes in:
-//  - The user's windowresolution = value (default, WxH, small, medium, large,
-//  desktop (or junk!)
-//  - The previously configured scaling mode (NONE, NEAREST, PERFECT)
-//  - If aspect correction (stretched pixels) is requested
+//  - The user's windowresolution: default, WxH, small, medium, large,
+//    desktop, or an invalid setting.
+//  - The previously configured scaling mode: NONE, NEAREST, or PERFECT.
+//  - If aspect correction (stretched pixels) is requested.
 
-// Except for SURFACE rendering, this function sets:
-//  - 'sdl.desktop.requested_window_bounds' with coarse bounds
-//  - 'sdl.desktop.window' with the refined size
+// Except for SURFACE rendering, this function returns a refined size and
+// additionally populates the following struct members:
+//  - 'sdl.desktop.requested_window_bounds', with the coarse bounds, which do
+//     not take into account scaling or aspect correction.
+//  - 'sdl.desktop.window', with the refined size.
+//  - 'sdl.desktop.want_resizable_window', if the window can be resized.
 
-// The coarse bounds do not take into account scaling or aspect correction,
-// while the refined size does (and because of this, is always a smaller
-// resolution than the coarse bounds).
+// Refined sizes:
+//  - If the user requested an exact resolution or 'desktop' size, then the
+//    requested resolution is only trimed for aspect (4:3 or 8:5), unless
+//    pixel-perfect is requested, in which case the resolution is adjusted down
+//    toward to nearest exact integer multiple.
 
-// We use the coarse bounds when re-calculating new sizes (such as
-// pixel-perfect) whenever there's a DOS-side resolution change.  This ensures
-// that the window will always be as big as possible, but still within bounds.
+//  - If the user requested a relative size: small, medium, or large, then either
+//    the closest fixed resolution is picked from a list or a pixel-perfect
+//    resolution is calculated. In both cases, aspect correction is factored in.
 
 static void setup_window_sizes_from_conf(const char *windowresolution_val,
                                          const SCALING_MODE scaling_mode,
@@ -2148,14 +2153,27 @@ static void setup_window_sizes_from_conf(const char *windowresolution_val,
 	assert(desktop.w >= FALLBACK_WINDOW_DIMENSIONS.x);
 	assert(desktop.h >= FALLBACK_WINDOW_DIMENSIONS.y);
 
-	// Bound the desktop resolution by the user's window-size setting
-	const std::string pref = NormalizeConfValue(windowresolution_val);
-	SDL_Point coarse_size;
-	if (pref.find('x') != std::string::npos)
-		coarse_size = window_bounds_from_resolution(pref, desktop);
-	else
-		coarse_size = window_bounds_from_label(pref, desktop);
+	// The refined scaling mode tell the refiner how it should adjust
+	// the coarse-resolution.
+	SCALING_MODE refined_scaling_mode = scaling_mode;
+	auto drop_nearest = [scaling_mode]() {
+		return (scaling_mode == SCALING_MODE::NEAREST) ? SCALING_MODE::NONE
+		                                               : scaling_mode;
+	};
 
+	// Get the coarse resolution from the users setting, and adjust
+	// refined scaling mode if an exact resolution is desired.
+	const std::string pref = NormalizeConfValue(windowresolution_val);
+	SDL_Point coarse_size = FALLBACK_WINDOW_DIMENSIONS;
+	if (pref.find('x') != std::string::npos) {
+		coarse_size = window_bounds_from_resolution(pref, desktop);
+		refined_scaling_mode = drop_nearest();
+	} else {
+		coarse_size = window_bounds_from_label(pref, desktop);
+		if (pref == "desktop") {
+			refined_scaling_mode = drop_nearest();
+		}
+	}
 	// Save the coarse bounds in the SDL struct for future sizing events
 	sdl.desktop.requested_window_bounds = {coarse_size.x, coarse_size.y};
 

--- a/src/gui/sdlmain.cpp
+++ b/src/gui/sdlmain.cpp
@@ -1995,8 +1995,8 @@ static SDL_Point remove_stretched_aspect(const SDL_Point &size)
 }
 
 static SDL_Point refine_window_size(const SDL_Point &size,
-                                              const SCALING_MODE scaling_mode,
-                                              const bool wants_stretched_pixels)
+                                    const SCALING_MODE scaling_mode,
+                                    const bool wants_stretched_pixels)
 {
 	switch (scaling_mode) {
 	case (SCALING_MODE::NONE): {
@@ -2079,13 +2079,18 @@ static SDL_Point window_bounds_from_resolution(const std::string &pref,
 	int w = 0;
 	int h = 0;
 	const bool was_parsed = sscanf(pref.c_str(), "%dx%d", &w, &h) == 2;
-	const bool is_valid = (w > 0 && w <= desktop.w && h > 0 && h <= desktop.h);
+
+	const bool is_out_of_bounds = (w > desktop.w || h > desktop.h);
+	if (was_parsed && is_out_of_bounds)
+		LOG_MSG("MAIN: Requested windowresolution '%dx%d' is larger than the desktop '%dx%d'",
+		        w, h, desktop.w, desktop.h);
+
+	const bool is_valid = (w > 0 && h > 0);
 	if (was_parsed && is_valid)
 		return {w, h};
 
-	LOG_MSG("MAIN: windowresolution '%s' outside of desktop '%dx%d', using %dx%d instead",
-	        pref.c_str(), desktop.w, desktop.h,
-	        FALLBACK_WINDOW_DIMENSIONS.x, FALLBACK_WINDOW_DIMENSIONS.y);
+	LOG_MSG("MAIN: Requested windowresolution '%s' is not valid, falling back to '%dx%d' instead",
+	        pref.c_str(), FALLBACK_WINDOW_DIMENSIONS.x, FALLBACK_WINDOW_DIMENSIONS.y);
 
 	return FALLBACK_WINDOW_DIMENSIONS;
 }
@@ -2103,7 +2108,7 @@ static SDL_Point window_bounds_from_label(const std::string &pref,
 	else if (pref == "desktop")
 		percent = 100;
 	else
-		LOG_MSG("MAIN: The windowresolution '%s' is invalid, using 'default' instead",
+		LOG_MSG("MAIN: Requested windowresolution '%s' is invalid, using 'default' instead",
 		        pref.c_str());
 
 	const int w = ceil_sdivide(desktop.w * percent, 100);
@@ -2177,11 +2182,9 @@ static void setup_window_sizes_from_conf(const char *windowresolution_val,
 	// Save the coarse bounds in the SDL struct for future sizing events
 	sdl.desktop.requested_window_bounds = {coarse_size.x, coarse_size.y};
 
-	// Refine the coarse size and save it in the SDL struct
-	const auto refined_size = refine_window_size_with_mode(coarse_size,
-	                                                       scaling_mode,
-	                                                       wants_stretched_pixels);
-
+	// Refine the coarse resolution and save it in the SDL struct
+	const auto refined_size = refine_window_size(coarse_size, refined_scaling_mode,
+	                                             wants_stretched_pixels);
 	assert(refined_size.x <= UINT16_MAX && refined_size.y <= UINT16_MAX);
 	sdl.desktop.window.width = static_cast<uint16_t>(refined_size.x);
 	sdl.desktop.window.height = static_cast<uint16_t>(refined_size.y);

--- a/src/gui/sdlmain.cpp
+++ b/src/gui/sdlmain.cpp
@@ -806,9 +806,9 @@ static SDL_Window *setup_window_pp(SCREEN_TYPES screen_type, bool resizable)
 	const int imgh = sdl.pp_scale.y * sdl.draw.height;
 
 	if (previous_imgw != imgw || previous_imgh != imgh)
-		LOG_MSG("MAIN: Pixel-perfect scaling (%dx%d): %dx%d (PAR %#.3g) -> %dx%d (PAR %#.3g)",
-		        sdl.pp_scale.x, sdl.pp_scale.y, sdl.draw.width,
-		        sdl.draw.height, sdl.draw.pixel_aspect, imgw, imgh,
+		LOG_MSG("MAIN: Scaling source %dx%d (PAR %#.3g) by %dx%d -> %dx%d (PAR %#.3g)",
+		        sdl.draw.width, sdl.draw.height, sdl.draw.pixel_aspect,
+		        sdl.pp_scale.x, sdl.pp_scale.y, imgw, imgh,
 		        static_cast<double>(sdl.pp_scale.y) / sdl.pp_scale.x);
 
 	const int wndw = (sdl.desktop.fullscreen ? w : imgw);
@@ -989,8 +989,7 @@ Bitu GFX_SetSize(Bitu width, Bitu height, Bitu flags,
 	                         sdl.draw.height != static_cast<int>(height) ||
 	                         sdl.draw.scalex != scalex ||
 	                         sdl.draw.scaley != scaley ||
-	                         sdl.draw.pixel_aspect != pixel_aspect ||
-	                         sdl.draw.callback != callback);
+	                         sdl.draw.pixel_aspect != pixel_aspect);
 
 	sdl.draw.width = static_cast<int>(width);
 	sdl.draw.height = static_cast<int>(height);
@@ -1003,7 +1002,7 @@ Bitu GFX_SetSize(Bitu width, Bitu height, Bitu flags,
 	const bool double_w = (flags & GFX_DBL_W) > 0;
 
 	if (is_changed)
-		LOG_MSG("MAIN: Draw resolution: %dx%d,%s%s pixel aspect ratio: %#.2f",
+		LOG_MSG("DISPLAY: Source resolution changed to %dx%d,%s%s (PAR %#.2f)",
 		        sdl.draw.width, sdl.draw.height,
 		        (double_w ? " double-width," : ""),
 		        (double_h ? " double-height," : ""), pixel_aspect);
@@ -2148,6 +2147,7 @@ static void setup_window_sizes_from_conf(const char *windowresolution_val,
 	if (sdl.desktop.want_type == SCREEN_SURFACE)
 		return;
 
+	// Can the window be resized?
 	sdl.desktop.want_resizable_window = detect_resizable_window();
 
 	// Get the total desktop resolution
@@ -2190,18 +2190,18 @@ static void setup_window_sizes_from_conf(const char *windowresolution_val,
 
 	// Let the user know the resulting window properties
 	if (scaling_mode == SCALING_MODE::NONE)
-		LOG_MSG("MAIN: Sized window to %dx%d with %s pixels",
-		        refined_size.x, refined_size.y,
+		LOG_MSG("MAIN: Sized window on display-%d to %dx%d with %s pixels",
+		        sdl.display_number, refined_size.x, refined_size.y,
 		        wants_stretched_pixels ? "stretched" : "square");
 
 	else if (scaling_mode == SCALING_MODE::NEAREST)
-		LOG_MSG("MAIN: Sized window to %dx%d with nearest-neighbour scaling and %s pixels",
-		        refined_size.x, refined_size.y,
+		LOG_MSG("MAIN: Sized window on display-%d to %dx%d with nearest-neighbour scaling and %s pixels",
+		        sdl.display_number, refined_size.x, refined_size.y,
 		        wants_stretched_pixels ? "stretched" : "square");
 
 	else if (scaling_mode == SCALING_MODE::PERFECT)
-		LOG_MSG("MAIN: Sized window %dx%d with pixel-perfect scaling and %s pixels",
-		        refined_size.x, refined_size.y,
+		LOG_MSG("MAIN: Sized window on display-%d to %dx%d with pixel-perfect scaling and %s pixels",
+		        sdl.display_number, refined_size.x, refined_size.y,
 		        wants_stretched_pixels ? "stretched" : "square");
 }
 

